pkg/erofs/erofs.go

```diff
@@ -0,0 +1,622 @@
+	Checksum           uint32
+	FeatureCompat      uint32
+	BlockSizeBits      uint8
+	SuperBlockExtSlots uint8
+	RootNid            uint16
+	Inodes             uint64
+	BuildTime          uint64
+	BuildTimeNsec      uint32
+	Blocks             uint32
+	MetaBlockAddr      uint32
+	XattrBlockAddr     uint32
+	UUID               [16]uint8
+	VolumeName         [16]uint8
+	FeatureIncompat    uint32
+	Union1             uint16
+	ExtraDevices       uint16
+	DevTableSlotOff    uint16
+	Reserved           [38]uint8
+}
+
+func (sb *SuperBlock) BlockSize() uint32 {
+	return 1 << sb.BlockSizeBits
+}
+
+func (sb *SuperBlock) BlockNumToAddr(nr uint32) uint64 {
+	return uint64(nr) << sb.BlockSizeBits
+}
+
+func (sb *SuperBlock) MetaOffset() uint64 {
+	return sb.BlockNumToAddr(sb.MetaBlockAddr)
+}
+
+func (sb *SuperBlock) NidToOffset(nid uint64) uint64 {
+	return sb.MetaOffset() + (nid << ISLOT_BITS)
+}
+
+type InodeCompact struct {
+	Format       uint16
+	XattrCount   uint16
+	Mode         uint16
+	Nlink        uint16
+	Size         uint32
+	Reserved     uint32
+	RawBlockAddr uint32
+	Ino          uint32
+	Uid          uint16
+	Gid          uint16
+	Reserved2    uint32
+}
+
+type InodeExtended struct {
+	Format       uint16
+	XattrCount   uint16
+	Mode         uint16
+	Reserved     uint16
+	Size         uint64
+	RawBlockAddr uint32
+	Ino          uint32
+	Uid          uint32
+	Gid          uint32
+	Mtime        uint64
+	MtimeNsec    uint32
+	Nlink        uint32
+	Reserved2    [16]uint8
+}
+
+type Dirent struct {
+	Nid      uint64
+	NameOff  uint16
+	FileType uint8
+	Reserved uint8
+}
+
+// +stateify savable
+type Image struct {
+	src   *os.File `state:"nosave"`
+	bytes []byte   `state:"nosave"`
+	sb    *SuperBlock
+	root  *Inode
+}
+
+func OpenImage(in *os.File) (*Image, error) {
+	i := &Image{src: in}
+	if sb, err := i.readSuperBlock(); err != nil {
+		return nil, err
+	} else {
+		i.sb = sb
+	}
+
+	if stat, err := i.src.Stat(); err != nil {
+		return nil, err
+	} else if bytes, err := syscall.Mmap(int(i.src.Fd()), 0, int(stat.Size()), syscall.PROT_READ, syscall.MAP_SHARED); err != nil {
+		return nil, err
+	} else {
+		i.bytes = bytes
+	}
+
+	if inode, err := i.newInode(FT_DIR, uint64(i.sb.RootNid)); err != nil {
+		syscall.Munmap(i.bytes)
+		return nil, err
+	} else {
+		i.root = inode
+	}
+	return i, nil
+}
+
+func (i *Image) Close() {
+	syscall.Munmap(i.bytes)
+	i.src.Close()
+}
+
+func (i *Image) BlockSize() uint32 {
+	return i.sb.BlockSize()
+}
+
+func (i *Image) Blocks() uint32 {
+	return i.sb.Blocks
+}
+
+func (i *Image) Root() *Inode {
+	return i.root
+}
+
+func (i *Image) readSuperBlock() (*SuperBlock, error) {
+	sb := &SuperBlock{}
+
+	if _, err := i.src.Seek(SUPER_BLOCK_OFFSET, os.SEEK_SET); err != nil {
+		return nil, fmt.Errorf("failed to reposition offset for super block: %v", err)
+	}
+
+	if err := binary.Read(i.src, binary.LittleEndian, sb); err != nil {
+		return nil, fmt.Errorf("failed to read super block: %v", err)
+	}
```

Comment:
Please use the go_marshal tool, it was built to replace `binary.Read` instances. It is much faster. It is code autogenerating tool we build to autogenerate fast code to do exactly this.

You need to do the following:
- Add `marshal = True` in the go_library definition for erofs in the BUILD file. 
- Annotate the SuperBlock struct with +marshal
```
// +marshal
type SuperBlock struct {
...
}
```
- This will autogenerate the [`Marshallable` interface](https://cs.opensource.google/gvisor/gvisor/+/master:pkg/marshal/marshal.go;l=51) for this struct.
- Use the `UnmarshalUnsafe()` method to deserialize from `i.bytes[SUPER_BLOCK_OFFSET:]`.

You can find a lot of usages of go_marshal: https://cs.opensource.google/search?q=%22%2F%2F%20%2Bmarshal%22&sq=&ss=gvisor%2Fgvisor